### PR TITLE
App authenticator docs and cli client

### PR DESCRIPTION
**Pull request by NikoHadouken**
* move app authenticator docs from readme to separate file
* deleted openapi.yaml
* updated docs
  * replaced current authentication mechanism with JWT signature tokens
  * renamed `device_id` to `authenticator_id`
* add authenticator cli

### CODE REVIEW

File: docs/App Authenticator.md [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+
+The API endpoints require an authentication mechanism that leverages client-side generated keypairs. While drafts for HTTP Message Signatures exist, they are no well-established standards.
+[draft-ietf-httpbis-message-signatures](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-message-signatures).
+
+Instead of implementing concepts from this unfinished draft, the API uses client-side generated JSON Web Tokens (JWTs) as a form of request signature. The JWTs are signed using a private key stored on the client and transmitted in the `x-signature` header.
+
+The client should use the authenticator id as the `kid` claim and use an asymmetric signature algorithm. The acceptable algorithms are:
+
+-   `PS512` with an `RSASSA-PSS` asymmetric key
+-   `ES512` with an `EC` asymmetric key
```
Comment (melegiul) [inline docs/App Authenticator.md:107]: I think thats fine. I want to stick close to the on board Keycloak JWT toolset and Keycloak seems to provide signature verifiers for your proposed algorithms: https://github.com/keycloak/keycloak/tree/main/services/src/main/java/org/keycloak/crypto

**Review by melegiul — CHANGES_REQUESTED** [review]
Thanks a lot for the CLI tool. It should work great despite I will need first to implement our new signature procedure to test it. :tada: 

I just found some minor issues, see below.

File: app-authenticator-cli/authenticator.js [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+		tabId,
+		key,
+		deviceOs,
+		devicePushId,
+		publicKey,
+		keyAlgorithm,
+	}) {
+		const jwt = await this._createSignatureToken({ typ: "app-setup-signature-token" });
+
+		await this._client.get("/login-actions/action-token", {
```
[the remark this answers is not in the document]
Comment (NikoHadouken) [reply app-authenticator-cli/authenticator.js:258]: The baseURL is set when creating a new KeycloakClient instance.
Here only the rest of the endpoint's path is required, not the baseURL.

File: docs/App Authenticator.md [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+
+##### ChallengeDTO
+
+```json
+{
+	"userName": "johndoe",
+	"userFirstName": "John",
+	"userLastName": "Doe",
+	"targetUrl": "http://192.168.2.127:8080/realms/dev/login-actions/action-token?key=eyJh...",
+	"secret": "FlJj9I4WoezeR3MN...",
```
[the remark this answers is not in the document]
Comment (NikoHadouken) [reply docs/App Authenticator.md:192]: Copy Paste mistake by me. Well spotted 👍🏼

File: docs/App Authenticator.md [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+| `device_push_id (optional)` | query | The platform specific ID to receive push notifications. For android this is the Firebase ID                                                          |
+
+##### Responses
+
+-   `2xx` Authenticator was successfully registered
+-   `400 Bad Request` Missing or invalid request parameters. This includes:
+    -   parsing of the JWT failed (invalid format)
+    -   request parameters are invalid (e.g. the signature or key algorithm is not supported)
+-   `401 Unauthorized` Verification of the JWT from the `key` query parameter failed in any form (expired, missing claims, invalid signature)
+-   `409 Precondition Failed` The authenticator ID is already registered
```
[the remark this answers is not in the document]
Comment (NikoHadouken) [reply docs/App Authenticator.md:155]: Mistake by me, meant to use 409 Conflict.

File: docs/App Authenticator.md [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+
+-   `2xx`challenge reply was successfully processed
+
+-   `400 Bad Request` Missing or invalid request parameters. This includes:
+    -   parsing of the JWT failed (invalid format)
+    -   the `x-signature` header has a wrong format
+-   `401 Unauthorized`
+    -   The `x-signature` header is missing or signature verification failed
+    -   The required JWT in query parameter `key` is expired or verification of the JWT failed in any other form (missing claims, invalid signature)
+-   `409 Precondition Failed` The referenced key (authenticator ID) in the request signature does not exist
```
[the remark this answers is not in the document]
Comment (NikoHadouken) [reply docs/App Authenticator.md:285]: In this case I wanted to use 412 Precondition Failed.
> The client has indicated preconditions in its headers which the server does not meet.

The precondition would be that the key is still registered on the server, but it got deleted by the user meanwhile.

File: app-authenticator-cli/authenticator.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (melegiul) [reply app-authenticator-cli/authenticator.js:258]: Sorry, you are right. My Bad.

File: docs/App Authenticator.md [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+| Name      | In   | Description           |
+| --------- | ---- | --------------------- |
+| `realmId` | path | The Keycloak realm ID |
+
+##### Responses
+
+-   `2xx` Login challenge dtos as array
+-   `400 Bad Request` the `x-signature` header has a wrong format
+-   `401 Unauthorized` The `x-signature` header is missing or verification failed
+-   `409 Precondition Failed` The referenced `kid` (authenticator ID) in the signature token does not exist
```
[the remark this answers is not in the document]
Comment (NikoHadouken) [reply docs/App Authenticator.md]: I meant 412 Precondition Failed for this one

[the remark this answers is not in the document]
Comment (NikoHadouken) [reply docs/App Authenticator.md]: We shouldn't overthink the status codes.
Choosing one where the description fits best our case and that enables us to distinguish error scenarios should be OK.
If you don't see another status code that fits better I would just stick with 412.

[the remark this answers is not in the document]
Comment (melegiul) [reply docs/App Authenticator.md]: Alright. Then lets take 412.

**Review by melegiul — APPROVED** [review]

**Pull request merged**